File: crawler.py
import time
import re
from selenium import webdriver
from datetime import datetime
from datetime import timedelta
from dateutil.parser import parse


class GoldenLeagueCrawler:

    def __init__(self, cookie_value, webdriver_path, binary_location='', retry_count=3):
        self.cookie_value = cookie_value
        self.webdriver_path = webdriver_path
        self.binary_location = binary_location
        self.retry_count = retry_count
        self.crawl_end = False

    def crawl(self):
        self.generate_selenium_driver()
        self.driver.get("https://p.eagate.573.jp/game/ddr/ddra20/")
        time.sleep(3)
        self.driver.add_cookie({'name': 'M573SSID', 'value': self.cookie_value})
        if not self.detail_crawl(self.retry_count):
            print("クロールに失敗しました。")
            self.driver.close()
            self.driver.quit()
            return False

        if self.start_date < self.now_datetime() <= self.end_date + timedelta(hours=1):
            self.gold_u_border, self.gold_l_border, self.time_text = self.border_crawl(self.ranking_page_url(self.league_count, 3), self.retry_count)
            self.silver_u_border, self.silver_l_border, self.time_text = self.border_crawl(self.ranking_page_url(self.league_count, 2), self.retry_count)
            self.bronze_u_border, self.bronze_l_border, self.time_text = self.border_crawl(self.ranking_page_url(self.league_count, 1), self.retry_count)
        
        self.driver.close()
        self.driver.quit()
        self.crawl_end = True
        return True

    def detail_crawl(self, retry_count):
        url = self.ranking_page_url()
        for _ in range(retry_count):
            self.driver.get(url)
            time.sleep(3)

            result = self.driver.find_elements_by_id("tournament")
            if len(result) == 0:
                continue

            self.league_count = int(re.sub(r"\D", "", result[0].text))
            dates_str = self.driver.find_elements_by_id("dates")[0].text
            self.start_date = parse(dates_str[5:21])
            self.end_date = parse(dates_str[22:])
            return True
        
        return False

    def border_crawl(self, url, retry_count):
        for i in range(retry_count):
            self.driver.get(url)
            time.sleep(3)
            idx = 0
            u_border = ""
            l_border = ""
            time_text = ""
            for td in self.driver.find_elements_by_xpath("//table[@class='table01'][2]/tbody/tr/td"):
                if idx == 0:
                    u_border = td.text
                if idx == 1:
                    l_border = td.text
                if idx == 2:
                    time_text = td.text
                idx += 1
            
            if u_border.isdecimal() or l_border.isdecimal():
                break
            
            if i == retry_count - 1:
                u_border = "取得に失敗しました。"
                l_border = "取得に失敗しました。"
        
        return u_border, l_border, time_text

    def tweet_text(self):
        if not self.crawl_end:
            return "クロールが終了していません。"

        return_text = ""
        if self.start_date == self.now_datetime():
            return "第%d回ゴールデンリーグ開始です。" % self.league_count
        if self.now_datetime() == self.end_date:
            return "第%d回ゴールデンリーグ終了です。現在集計中になります。" % self.league_count
        if self.now_datetime() == self.end_date + timedelta(hours=1):
            return_text = "第%d回ゴールデンリーグお疲れ様でした。\n最終結果\n\n" % self.league_count
        if self.start_date < self.now_datetime() < self.end_date:
            return_text = self.time_text + " 時点のボーダー:\n\n"
        if self.now_datetime() < self.start_date or self.end_date + timedelta(hours=1) < self.now_datetime():
            return ""
        
        return_text += "ゴールドクラス:\n"
        return_text += "降格ボーダー: " + self.gold_l_border + "\n\n"
        return_text += "シルバークラス:\n"
        return_text += "昇格ボーダー: " + self.silver_u_border + "\n"
        return_text += "降格ボーダー: " + self.silver_l_border + "\n\n"
        return_text += "ブロンズクラス:\n"
        return_text += "昇格ボーダー: " + self.bronze_u_border
        
        return return_text

    def generate_selenium_driver(self):
        options = webdriver.ChromeOptions()
        if not self.binary_location == '':
            options.binary_location = self.binary_location

        options.add_argument("--headless")
        options.add_argument("--disable-gpu")
        options.add_argument("--window-size=1280x1696")
        options.add_argument("--disable-application-cache")
        options.add_argument("--disable-infobars")
        options.add_argument("--no-sandbox")
        options.add_argument("--hide-scrollbars")
        options.add_argument("--enable-logging")
        options.add_argument("--log-level=0")
        options.add_argument("--v=99")
        options.add_argument("--single-process")
        options.add_argument("--ignore-certificate-errors")
        options.add_argument("--homedir=/tmp")

        self.driver = webdriver.Chrome(self.webdriver_path, options=options)

    def ranking_page_url(self, league_id=0, class_id=0):
        url = "https://p.eagate.573.jp/game/ddr/ddra20/p/ranking/index.html"
        if 0 < league_id and 0 < class_id:
            url += "?league_id=%d&class_id=%d" % league_id, class_id
        
        return url

    def now_datetime(self):
        if not hasattr(self, "now_time"):
            self.now_time = datetime(datetime.now().year, datetime.now().month, datetime.now().day, datetime.now().hour)

        return self.now_time

File: crawler.rb
require 'date'
require 'selenium-webdriver'

class GoldenLeagueCrawler
  SELENIUM_OPTIONS = %w[--headless --disable-gpu --disable-extensions --disable-dev-shm-usage --window-size=1280x1696
                        --disable-application-cache --disable-infobars --no-sandbox --hide-scrollbars --single-process
                        --ignore-certificate-errors --homedir=/tmp].freeze

  # コンストラクタ
  # @param [String] konami_cookie_str KONAMIのWebページで使用しているM573SSIDに対応するcookieの値
  # @param [String] chromium_path 使用するHeadless-ChromiumのPATH
  # @param [String] webdriver_path 使用するHeadless-Chromiumに対応しているWebDriverのPATH
  # @param [Integer] retry_count クロール失敗時にリトライする回数
  def initialize(konami_cookie_str, chromium_path, webdriver_path, retry_count)
    @konami_cookie_str = konami_cookie_str
    @chromium_path = chromium_path
    @webdriver_path = webdriver_path
    @retry_count = retry_count

    generate_selenium_driver
  end

  # 必要な箇所までのクロール処理を行う。
  def run
    set_konami_cookie
    raise 'ゴールデンリーグのページに接続失敗しました。' unless crawl(ranking_page_url)

    return true unless now? || result_now?

    gold_class_crawl
    silver_class_crawl
    bronze_class_crawl
    @last_update = last_update
    @driver.quit
  end

  # クロール結果を元にツイート内容を返す。
  # @return [String] ツイート内容
  # @return [nil] ゴールデンリーグ開催期間外
  def tweet_text
    return start_now_tweet if start_now?

    return now_tweet if now?

    return end_now_tweet if end_now?

    return result_now_tweet if result_now?

    nil
  end

  private

  # urlのページをクロールする。
  # @param [String] url クロール対象のURL
  # @return [Boolean] クロール成功した場合はtrue, 失敗した場合はfalse
  def crawl(url)
    count = 0

    begin
      sleep 2
      @driver.get(url)
    rescue
      count += 1 if count <= @retry_count
      retry if count <= @retry_count
    end

    count <= @retry_count
  end

  # DDRのWebページにアクセスし、コナミアカウントのcookieを設定する。
  def set_konami_cookie
    raise 'DDRのWebページに接続失敗しました。' unless crawl(ddr_page_url)

    @driver.manage.add_cookie('name': 'M573SSID', 'value': @konami_cookie_str)
  end

  # ゴールドクラスの降格ボーダーをクロールしインスタンス変数に保存する。
  def gold_class_crawl
    raise 'ゴールドクラスのページに接続失敗しました。' unless crawl(ranking_page_url(league_count, 3))

    @gold_lower = lower_border
  end

  # シルバークラスの昇格/降格ボーダーをクロールしインスタンス変数に保存する。
  def silver_class_crawl
    raise 'シルバークラスのページに接続失敗しました。' unless crawl(ranking_page_url(league_count, 2))

    @silver_upper = upper_border
    @silver_lower = lower_border
  end

  # ブロンズクラスの昇格ボーダーをクロールしインスタンス変数に保存する。
  def bronze_class_crawl
    raise 'ブロンズクラスのページに接続失敗しました。' unless crawl(ranking_page_url(league_count, 1))

    @bronze_upper = upper_border
  end

  # 最新のゴールデンリーグの開始日時を返す。
  # @return [DateTime] ゴールデンリーグの開始日時
  def start_time
    @start_time ||= DateTime.parse(@driver.find_element(:id, 'dates').text[5..20])
  end

  # 最新のゴールデンリーグの終了日時を返す。
  # @return [DateTime] ゴールデンリーグの終了日時
  def end_time
    @end_time ||= DateTime.parse(@driver.find_element(:id, 'dates').text[22..37])
  end

  # 最後に開催されたリーグID または 開催中のリーグIDを返す。
  # @return [Integer] ゴールデンリーグのID
  def league_count
    @league_count ||= @driver.find_element(:id, 'tournament').text.gsub(/[^\d]/, '').to_i
  end

  # 現在表示しているページの昇格ボーダーの値を抜き出して返す。
  # @return [String] 現在表示しているページの昇格ボーダー
  def upper_border
    @driver.find_element(:xpath, '//table[@class="table01"][2]/tbody/tr[1]/td/p').text
  end

  # 現在表示しているページの降格ボーダーの値を抜き出して返す。
  # @return [String] 現在表示しているページの降格ボーダー
  def lower_border
    @driver.find_element(:xpath, '//table[@class="table01"][2]/tbody/tr[2]/td/p').text
  end

  # 現在表示しているページの集計日時の値を抜き出して返す。
  # @return [String] 現在表示しているページの集計日時
  def last_update
    @driver.find_element(:xpath, '//table[@class="table01"][2]/tbody/tr[3]/td/p').text
  end

  # ゴールデンリーグのランキング表示中かどうかを返す。
  # @return [Boolean] ゴールデンリーグ開催中ならtrue, 非開催中ならfalse
  def now?
    now_date = DateTime.parse(DateTime.now.strftime('%Y-%m-%d %H'))
    start_time < now_date && now_date < end_time
  end

  # ゴールデンリーグ開始直後かどうかを返す。
  # @return [Boolean] ゴールデンリーグ開始直後ならtrue, その他はfalse
  def start_now?
    start_time == DateTime.parse(DateTime.now.strftime('%Y-%m-%d %H'))
  end

  # ゴールデンリーグ終了直後かどうかを返す。
  # @return [Boolean] ゴールデンリーグ終了直後ならtrue, その他はfalse
  def end_now?
    end_time == DateTime.parse(DateTime.now.strftime('%Y-%m-%d %H'))
  end

  # 最終結果発表かどうかを返す。
  # @return [Boolean] ゴールデンリーグ集計終了直後ならtrue, その他はfalse
  def result_now?
    now = DateTime.parse(DateTime.now.strftime('%Y-%m-%d %H'))
    now.year == end_time.year && now.month == end_time.month && now.day == end_time.day && now.hour + 1 == end_time.hour
  end

  # DDRのゴールデンリーグのページのURLを返す。
  # league_idで指定したリーグとclass_idで指定したクラス(gold:3, silver:2, bronze:1)のページのURLを返す。
  # 指定しない場合は最新のリーグの所属リーグのページのURLを返す。
  #
  # @param [Integer] league_id 何回目のリーグかを示すリーグID
  # @param [Integer] class_id クラス毎に指定されている値(gold:3, silver:2, bronze:1)
  # @return [String] ゴールデンリーグのページのURL
  def ranking_page_url(league_id = -1, class_id = -1)
    url = "#{ddr_page_url}/ranking/index.html"
    url + (0 <= league_id && 0 <= class_id ? "?league_id=#{league_id}&class_id=#{class_id}" : '')
  end

  # DDRのページのURLを返す
  # @return [String] DDRのページのURL
  def ddr_page_url
    'https://p.eagate.573.jp/game/ddr/ddra20/p'
  end

  # headless-chromeのWebDriverインスタンスを生成して@driver変数に保存する。
  def generate_selenium_driver
    Selenium::WebDriver::Chrome::Service.driver_path = @webdriver_path
    options = Selenium::WebDriver::Chrome::Options.new(binary: @chromium_path)

    SELENIUM_OPTIONS.each { |option| options.add_argument(option) }
    @driver = Selenium::WebDriver.for :chrome, options: options
  end

  # ゴールデンリーグ開始直後のツイート内容を返す。
  # @return [String] ツイート内容
  def start_now_tweet
    "第#{league_count}回ゴールデンリーグ開始です。"
  end

  # ゴールデンリーグ終了直後のツイート内容を返す。
  # @return [String] ツイート内容
  def end_now_tweet
    "第#{league_count}回ゴールデンリーグ終了です。現在集計中になります。"
  end

  # ゴールデンリーグ集計完了直後のツイート内容を返す。
  # @return [String] ツイート内容
  def result_now_tweet
    "第#{league_count}回ゴールデンリーグお疲れ様でした。\n最終結果\n\n#{score_text}"
  end

  # ゴールデンリーグ中のツイート内容を返す。
  # @return [String] ツイート内容
  def now_tweet
    "#{@last_update} 時点のボーダー:\n\n#{score_text}"
  end

  def score_text
    "ゴールドクラス:\n降格ボーダー: #{@gold_lower}\n\n" \
    "シルバークラス:\n昇格ボーダー: #{@silver_upper}\n降格ボーダー: #{@silver_lower}\n\n" \
    "ブロンズクラス:\n昇格ボーダー: #{@bronze_upper}\n"
  end
end
